### DISCUSSION TITLE
Minimal number of replicas

### DISCUSSION THREAD

**Discussion by atanev** [cloudnative-pg/cloudnative-pg#3638] [Q&A] [answered]
Hello!
Thank you for the amazing project and documenting so much of it allowing me to (hopefully) ask a better educated question.

I have a use case where the number of Kubernetes nodes is limited but I still would like to run a Highly Available Postgres cluster. Ideally I would prefer to run 1 primary and 1 standby (read-only) with replication is case of a failure on the Primary`s node.

**disclaimer**: below is my interpretation based on reading docs and might not reflect 100% the truth thus I`m opening this discussion here.
Most other operators use more traditional software for clustering like `patroni` and I`m uncertain how well such solutions can cope with potential split-brain scenario in a 2 instance setup (1 primary and 1 standby ro replica).
Reading through the documentation of CloudNativePG my understanding is the operator is handling the fail-over in case of a primary failure. ( that comes at a price of the Operator must not not be running on the same node as the primary so it can detect it )

Reading throughout the docs trying to back that up with data I came across this section part of F.A.Q:

```
Use at least one standby, preferably at least two, so that you can configure synchronous replication in the cluster, introducing RPO=0 for high availability.
```

Another ref: https://cloudnative-pg.io/documentation/current/replication/#select-nodes-for-synchronous-replication
```
As an example use-case for this feature: in a cluster with a single sync replica,....
```

Ofc going through the reference best practice architectures having 3+ is better but my question is:
Is 1 primary + 1 replica standby a viable solution that does not pose technology limitations that will prevent the fail-over due to the even number of nodes?

The architecture will be similar to the `shared nothing` architecture shown here: https://cloudnative-pg.io/documentation/current/architecture/#postgresql-architecture but instead of 2 only 1 hot standby replica.

Measures will be taken to place primary and standby to be placed on different nodes + separation from the CN-PG.

Thank you for your time and happy holidays.

**Comment by gbartolini** [comment, marked as answer]
Thanks for your nice words.

Short answer: 1+1 architecture does not prevent failover.

If you have reliable hardware underneath and you have tested thoroughly the resilience and reliability of the storage, you can live with a single replica. However, in order to achieve RPO=0 (so ensure that the replica has the same data as the primary), you need synchronous replication. Normally in Postgres outside Kubernetes you need at least two replicas (or a replica and a Barman server with WAL streaming, for example). This is because if you have sync replication, and no replica is available, writes hang until a replica can receive them.

CloudNativePG is smart to disable syncrep in case no replica is available, but it might generate unexpected data loss. For this reason we recommend having 2 replicas if you have RPO=0 requirements, at least until we can configure this behaviour (by disabling it for example).

So, to come to your question, 1+1 is OK if you don't have RPO=0 requirements. There's no quorum implication here (as the controller is Kubernetes itself which - hopefully - has a distributed control plane). The failover will work fine, provided that the replica is up (that's why shared nothing architectures are recommended).

I hope this helps. My advice is to read this article that I wrote a few months ago: https://www.cncf.io/blog/2023/09/29/recommended-architectures-for-postgresql-in-kubernetes/
